<h1>用户行为分析<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#分析背景" data-toc-modified-id="分析背景-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>分析背景</a></span><ul class="toc-item"><li><span><a href="#数据源" data-toc-modified-id="数据源-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>数据源</a></span></li><li><span><a href="#数据解释" data-toc-modified-id="数据解释-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>数据解释</a></span></li><li><span><a href="#分析目的" data-toc-modified-id="分析目的-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>分析目的</a></span></li><li><span><a href="#数据处理" data-toc-modified-id="数据处理-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>数据处理</a></span></li></ul></li><li><span><a href="#结论前置" data-toc-modified-id="结论前置-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>结论前置</a></span></li><li><span><a href="#整体情况" data-toc-modified-id="整体情况-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>整体情况</a></span><ul class="toc-item"><li><span><a href="#整体指标" data-toc-modified-id="整体指标-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>整体指标</a></span><ul class="toc-item"><li><span><a href="#流量指标" data-toc-modified-id="流量指标-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>流量指标</a></span></li><li><span><a href="#行为指标" data-toc-modified-id="行为指标-3.1.2"><span class="toc-item-num">3.1.2&nbsp;&nbsp;</span>行为指标</a></span></li><li><span><a href="#交易指标" data-toc-modified-id="交易指标-3.1.3"><span class="toc-item-num">3.1.3&nbsp;&nbsp;</span>交易指标</a></span></li></ul></li><li><span><a href="#整体转化漏斗" data-toc-modified-id="整体转化漏斗-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>整体转化漏斗</a></span></li></ul></li><li><span><a href="#时间维度分析" data-toc-modified-id="时间维度分析-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>时间维度分析</a></span><ul class="toc-item"><li><span><a href="#按周分析" data-toc-modified-id="按周分析-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>按周分析</a></span></li><li><span><a href="#按星期分析" data-toc-modified-id="按星期分析-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>按星期分析</a></span></li><li><span><a href="#按日分析" data-toc-modified-id="按日分析-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>按日分析</a></span></li><li><span><a href="#按时段分析" data-toc-modified-id="按时段分析-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>按时段分析</a></span></li></ul></li><li><span><a href="#用户分析" data-toc-modified-id="用户分析-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>用户分析</a></span><ul class="toc-item"><li><span><a href="#用户行为统计信息" data-toc-modified-id="用户行为统计信息-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>用户行为统计信息</a></span></li><li><span><a href="#用户活跃" data-toc-modified-id="用户活跃-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>用户活跃</a></span></li><li><span><a href="#用户留存" data-toc-modified-id="用户留存-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>用户留存</a></span></li><li><span><a href="#用户登录天数" data-toc-modified-id="用户登录天数-5.4"><span class="toc-item-num">5.4&nbsp;&nbsp;</span>用户登录天数</a></span></li><li><span><a href="#用户购买情况" data-toc-modified-id="用户购买情况-5.5"><span class="toc-item-num">5.5&nbsp;&nbsp;</span>用户购买情况</a></span></li><li><span><a href="#行为路径分析" data-toc-modified-id="行为路径分析-5.6"><span class="toc-item-num">5.6&nbsp;&nbsp;</span>行为路径分析</a></span></li></ul></li><li><span><a href="#客户分析" data-toc-modified-id="客户分析-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>客户分析</a></span></li><li><span><a href="#品牌分析" data-toc-modified-id="品牌分析-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>品牌分析</a></span></li></ul></div>

*Author: json.wong.work@gmail.com*

*Index: https://jsonww.gitee.io/pdf/movielens.pdf*

## 分析背景


### 数据源

数据源：[eCommerce Events History in Cosmetics Shop | Kaggle](https://www.kaggle.com/mkechinov/ecommerce-events-history-in-cosmetics-shop?select=2020-Feb.csv)

数据文件：
- **2020-Feb.csv**：某化妆品电商网站 11 月份用户行为数据。

### 数据解释

|字段|数据类型|解释|
|-|-|-|
|event_type|字符串|行为类型：{view(浏览)，cart(加入购物车)，purchase(购买)，remove_from_cart(移出购物车)}|
|event_time|datetime|行为发生时间|
|user_id|非负整数|用户 id|
|product_id|非负整数|商品 id|
|category_id|非负整数|分类 id|
|brand|字符串|品牌名称|
|price|正实数|商品价格|

### 分析目的

了解运营情况，用户分层，品牌分析，调整策略，提出建议。

### 数据处理

数据已经过 BehaviourAnalysisPreprocessing 模块预处理。本次分析前进行简单再处理如下：

- event_type：一个行为事件对应一次行为，故转为独热编码
- event_time：此次分析时间粒度不需要到分秒级，故只保留月、周天、日期、时段

In [1]:
import numpy as np
import pandas as pd
pd.set_option('display.float_format',lambda x : '%.6f' % x)
pd.set_option('display.max_columns', 32)

import matplotlib.pyplot as plt
import seaborn as sns
sns.set_style('whitegrid')
from pyecharts.charts import *
from pyecharts.components import Table
from pyecharts.commons.utils import JsCode
import pyecharts.options as opts
from pyecharts.globals import ThemeType

import os
import warnings
warnings.filterwarnings('ignore')


"""基础环境配置"""
DATA_PATH = '../data/'
CLEAN_PATH = '../clean/'
CACHE_PATH = '../cache/'
SUBMIT_PATH = '../submit/'
MODEL_PATH = '../model/'
DATA_FILE = "2019-Nov.csv"
cleaned_target = CLEAN_PATH + f"processed_data.h5"
"""主题配置"""
# PYECHART_THEME = 

def reduce_memory(df, exclude_columns=[], detial=False):
    '''Adjust data dtype to save memory.'''
    start_mem = df.memory_usage().sum()    
    reduce_columns = [col for col in df.columns if col not in exclude_columns]
    
    for col in reduce_columns:
        col_type = df[col].dtype
        
        if object != col_type:
            col_min, col_max = df[col].min(), df[col].max()
            if 'int' == str(col_type)[:3]:
                if col_min > np.iinfo(np.int8).min and col_max < np.iinfo(np.int8).max:
                    df[col] = df[col].astype(np.int8)
                elif col_min > np.iinfo(np.int16).min and col_max < np.iinfo(np.int16).max:
                    df[col] = df[col].astype(np.int16)
                elif col_min > np.iinfo(np.int32).min and col_max < np.iinfo(np.int32).max:
                    df[col] = df[col].astype(np.int32)
                elif col_min > np.iinfo(np.int64).min and col_max < np.iinfo(np.int64).max:
                    df[col] = df[col].astype(np.int64)  
            else:
                if col_min > np.finfo(np.float16).min and col_max < np.finfo(np.float16).max:
                    df[col] = df[col].astype(np.float16)
                elif col_min > np.finfo(np.float32).min and col_max < np.finfo(np.float32).max:
                    df[col] = df[col].astype(np.float32)
                else:
                    df[col] = df[col].astype(np.float64)
        else:
            df[col] = df[col].astype('category')

    if detial:
        end_mem = df.memory_usage().sum()
        print(f"Memory before reduce: {start_mem:.2f}")
        print(f"Memory after reduce: {end_mem:.2f}")
        print(f"Decreased by: {(start_mem - end_mem) / start_mem:.2%}")
    return df


# 读取清洗后的数据，并进行数据类型调整以减少内存占用：
data = pd.read_hdf(cleaned_target, key='processed')
data = pd.concat([data, pd.get_dummies(data['event_type'])], axis=1)
# data = data.drop(columns=['event_type'])  # 留作路径聚合
data['event_time'] = pd.to_datetime(data['event_time'].apply(lambda x: x[:-4]))
data['month'] = data['event_time'].dt.month
data['week'] = data['event_time'].dt.week
data['weekday'] = data['event_time'].dt.weekday
data['day'] = data['event_time'].dt.day
data['hour'] = data['event_time'].dt.hour
data = data.drop(columns=['event_time'])
data = reduce_memory(data, ['price', 'cart', 'purchase', 'remove_from_cart', 'view', 'turnovers'])
# data.info()

## 结论前置

## 整体情况

1. 现象：用户人均浏览次数 5.63，跳失率为 45.54%。

    结论：跳失率较高，需要降低跳失率。

    建议：
    - 改善页面设计和产品展示，提高用户体验，提升用户第一印象，以促进用户进一步操作。
    - 对老用户加强定制化推荐，提高用户浏览的兴趣，促进用户进一步的操作。
    - 在用户浏览过程中创建场景，诱导产生进一步行为。如： 提示用户加购/关注可即时获得价格变动、活动信息。

### 整体指标

#### 流量指标

指标口径：
- pv：访问量
- uv：独立访客数
- 跳失率：仅有“1 次访问行为”的用户比例
- 人均浏览次数：平均每个用户浏览次数

In [2]:
pv = data['view'].sum()
uv = data['user_id'].nunique()
user_behaviour_gb = data[['user_id', 'view', 'cart', 'remove_from_cart', 'purchase']].groupby('user_id').sum()
bounce_rate = len(user_behaviour_gb[(user_behaviour_gb['view'] == 1) & \
                                    (user_behaviour_gb['cart'] + user_behaviour_gb['remove_from_cart'] + user_behaviour_gb['purchase'] == 0)]) / len(user_behaviour_gb)
per_capita_view = user_behaviour_gb['view'].mean()

headers = ['PV', 'UV', '跳失率', '人均浏览次数']
flow= [[pv, uv, f"{bounce_rate:.2%}", f"{per_capita_view:.2f}"]]
flow_table = (
    Table()
    .add(headers, flow)
)
flow_table.render_notebook()
# page = Page()
# page.add(flow_table, behaviour_sum_table, transaction_table)
# page.render_notebook()

现象：用户人均浏览次数 5.63，跳失率为 45.54%。

结论：跳失率较高，需要降低跳失率。

建议：
- 改善页面设计和产品展示，提高用户体验，提升用户第一印象，以促进用户进一步操作。
- 对老用户加强定制化推荐，提高用户浏览的兴趣，促进用户进一步的操作。
- 在用户浏览过程中创建场景，诱导产生进一步行为。如： 提示用户加购/关注可即时获得价格变动、活动信息。

#### 行为指标

指标口径：
- 加购次数：'cart'次数
- 移出购物车次数：'remove_from_cart'次数
- 购买次数：'purchase'次数
- 月平均使用天数：当月每天登陆用户数（去重）之和与当月登陆用户数（去重）的比值

In [3]:
cart_n, rmc_n, purchase_n = data['cart'].sum(), data['remove_from_cart'].sum(), data['purchase'].sum()
month_average_login = len(data[['user_id', 'day']]) / data['user_id'].nunique()

headers = ['加购次数', '移出购物车次数', '购买次数', '月平均使用天数']
behaviour_sum = [[cart_n, rmc_n, purchase_n, f"{month_average_login:.0f}"]]
behaviour_sum_table = (
    Table()
    .add(headers, behaviour_sum)
)
behaviour_sum_table.render_notebook()

#### 交易指标

指标口径：
- 成交额：购买物品的价格总数
- 客户数：有购买记录的用户数
- 客单价：成交额除以购买客户数
- 购买率：所有用户中，有购买记录的用户占比
- 复购率：有购买记录的用户中，有两次以上购买记录的用户占比
- 平均购买次数：购买用户的平均购买次数

In [4]:
turnovers = data['turnovers'].sum()
user_gb = data[['user_id', 'view', 'cart', 'remove_from_cart', 'purchase', 'turnovers']].groupby('user_id').sum()
buyer_gb = user_gb[user_gb['purchase'] > 0]
customer_unit_price = turnovers / buyer_gb.shape[0]
purchase_rate = len(buyer_gb) / len(user_gb)
repurchase_rate = len(buyer_gb[buyer_gb['purchase'] > 1]) / len(buyer_gb)

headers = ['成交额', '消费用户数', '客单价', '购买率', '复购率', '平均购买次数']
transaction = [[
    f"{turnovers:.2f}",
    f"{buyer_gb.shape[0]}",
    f"{customer_unit_price:.2f}",
    f"{purchase_rate:.2%}",
    f"{repurchase_rate:.2%}",
    f"{buyer_gb['purchase'].mean():.0f}"
]]
transaction_table = (
    Table()
    .add(headers, transaction)
)
transaction_table.render_notebook()

In [5]:
# all_pcb = data[['product_id', 'category_id', 'brand']].nunique()
# buy_pcb = data[data['purchase'] > 0][['product_id', 'category_id', 'brand']].nunique()
# buy_pcb / all_pcb

现象：
- 用户购买率仅有 8.56%。
- 客户复购率高达 91.42%。

结论：
- 用户购买率很低，新用户的运营亟待改善。
- 老用户运营状况良好，客户粘性很高。

建议：
- 加强对新用户的运营
    - 挖掘用户需求。如：优化推荐系统，提高商品和用户需求的匹配度。
    - 适当让利，诱导用户购买。如：首单优惠。
- 调整运营重心，权衡新用户转化和老用户复购转化的成本，在尽量保持老用户复购转化的情况下，加强对新用户的运营。
- 发挥老用户忠诚度高的优势，尝试以老带新的运营方法。如：邀请码注册优惠。

### 整体转化漏斗

In [6]:
def draw_behaviour_funnel(data, behaviour_lst, overall=True, is_show=True, title='Funnel'):
    behaviour_df = pd.DataFrame(data[behaviour_lst].sum(), index=behaviour_lst, columns=['n'])
    if overall:
        behaviour_df['转化率'] = behaviour_df['n'] / behaviour_df.iloc[0, 0]
    else:
        behaviour_df['转化率'] = behaviour_df['n'].shift(-1) / behaviour_df['n']
        behaviour_df['转化率'] = behaviour_df['转化率'].shift(1)
        behaviour_df.iloc[0, 1] = 1
    behaviour_df['转化率'] = behaviour_df['转化率'].apply(lambda x: format(x, ".2%"))
    behaviour_df['转化环节'] = behaviour_df.index + ': ' + behaviour_df['转化率']
#     print(pd.DataFrame(behaviour_df))

    behaviour_funnel = (
        Funnel()
        .add(
            '',
            behaviour_df[['转化环节', 'n']].values,
        )
        .set_global_opts(
            title_opts=opts.TitleOpts(
                title=title,
                subtitle='转化率计算方式：每个环节以第一个环节作为计算基准' if overall else '计算方式：每个环节以前一个环节作为计算基准',
            ),
            legend_opts=opts.LegendOpts(
                pos_left='right',
                pos_top='bottom',
                orient='vertical'
            ),
        )
        .set_series_opts(
            label_opts=opts.LabelOpts(
                position='inner',
                formatter='{b}',
            )
        )
    )
    return behaviour_funnel.render_notebook() if is_show else behaviour_funnel

In [7]:
title = '浏览-加购-购买整体转化'
draw_behaviour_funnel(data, ['view', 'cart', 'purchase'], title=title)

In [8]:
title = '浏览-加购-移出购物车整体转化'
draw_behaviour_funnel(data, ['view', 'cart', 'remove_from_cart'], title=title)

现象：“浏览-加购”整体转化率为 62.12%，而“加购-移出购物车”的整体转换率为 33.76%，是“加购-购买”的整体转化率 15.53% 的两倍。

结论：
- “浏览-加购”的转化率较高，说明用户善用加购功能。
- “加购-移出购物车”的比例明显高于“加购-购买”，说明用户加购的目的并不只为了购买，还可能是商品收藏或价格观察。

建议：
- 推出收藏功能，既完善用户使用的功能性，又便于我们分析用户的意图。
- 适时调整商品价格，促进用户购买。例如：发放优惠券促销、清仓打折。
- 当加购后的商品发生信息变更或久未操作，及时提醒用户。例如：客户端-推送、web 端-消息通知。

## 时间维度分析

### 按周分析

In [9]:
def show_time_trend(uv_df, behaviour_df, bar_title='UV', bar_subtitle='-', line_title='Behaviour', line_subtitle='-', datazoom_opts_show=False, is_show=True):
    """Show the data trend of specific timescales."""
    
    # bar
    uv_bar = (
        Bar()
        .set_global_opts(
            title_opts=opts.TitleOpts(
                title=bar_title,
                subtitle=bar_subtitle,
            ),
            tooltip_opts=opts.TooltipOpts(
                is_show=True, trigger="axis", axis_pointer_type="cross"
            ),
            xaxis_opts=opts.AxisOpts(
                type_="category",
                axispointer_opts=opts.AxisPointerOpts(is_show=True, type_="shadow"),
            ),
        )
        .add_xaxis(uv_df.index.astype('str').tolist())
    )
    for c in uv_df.columns:
        uv_bar.add_yaxis(c, uv_df[c].tolist())
        uv_bar.set_series_opts(
            label_opts=opts.LabelOpts(is_show=False),
            markpoint_opts=opts.MarkPointOpts(
                data=[
                    opts.MarkPointItem(type_="max", name="最大值"),
                    opts.MarkPointItem(type_="min", name="最小值"),
                ]
            ),
            markline_opts=opts.MarkLineOpts(
                data=[
                    opts.MarkLineItem(name=f"{c}均值", type_="average")
                ],
                symbol='none',
                label_opts=opts.LabelOpts(is_show=False),
                linestyle_opts=opts.LineStyleOpts(
                    opacity=0.5,
                    type_='dashed'
                ),
            ),
        )
    if datazoom_opts_show:
        uv_bar.set_global_opts(datazoom_opts=opts.DataZoomOpts(is_show=datazoom_opts_show))
    
    # line
    behaviour_line = (
        Line()
        .set_global_opts(
            title_opts=opts.TitleOpts(
                title=line_title,
                subtitle=line_subtitle,
            ),
            tooltip_opts=opts.TooltipOpts(
                is_show=True, trigger="axis", axis_pointer_type="cross"
            ),
        )
        .add_xaxis(behaviour_df.index.astype(str).tolist())
    )
    for b in behaviour_df.columns:
        behaviour_line.add_yaxis(b, behaviour_df[b].tolist())
        behaviour_line.set_series_opts(
            label_opts=opts.LabelOpts(is_show=False),
            linestyle_opts=opts.LineStyleOpts(width=2),
            markpoint_opts=opts.MarkPointOpts(
                data=[
                    opts.MarkPointItem(type_="max", name="最大值"),
                    opts.MarkPointItem(type_="min", name="最小值"),
                ]
            ),
            markline_opts=opts.MarkLineOpts(
                data=[
                    opts.MarkLineItem(name=f"{b}均值", type_="average")
                ],
                symbol='none',
#                 label_opts=opts.LabelOpts(formatter=f"{b}均值"),
                label_opts=opts.LabelOpts(is_show=False),
                linestyle_opts=opts.LineStyleOpts(
                    opacity=0.5,
                    type_='dashed'
                ),
            ),
        )

    page = (
        Page()
            .add(uv_bar, behaviour_line)
    )
    return page.render_notebook() if is_show else page


def double_y(data, left, right, title='Double y-axis', left_kind=Line(), right_kind=Line(), is_show=True):
    """双 y 轴"""
    left_chart = (
        Line()
        .add_xaxis(data.index.astype('str').tolist())
        .set_global_opts(
            title_opts=opts.TitleOpts(
                title=title,
            ),
            tooltip_opts=opts.TooltipOpts(
                is_show=True, trigger="axis", axis_pointer_type="cross"
            ),
            xaxis_opts=opts.AxisOpts(
                name=f"{data.index.name}",
                name_location='middle',
                name_gap=30,
                type_="category",
                axispointer_opts=opts.AxisPointerOpts(is_show=True, type_="shadow"),
            ),
            yaxis_opts=opts.AxisOpts(
                name=f"{'/'.join(left)}",
            ),
        )
        .extend_axis(
            yaxis=opts.AxisOpts(
                name=f"{'/'.join(right)}",
                axislabel_opts=opts.LabelOpts(formatter="{value}"),
            )
        )
    )

    for l in left:
        left_chart.add_yaxis(
            l,
            data[l].tolist(),
            label_opts=opts.LabelOpts(is_show=False),
            linestyle_opts=opts.LineStyleOpts(width=2),
        )

    right_chart = (
        Line()
        .add_xaxis(data.index.astype('str').tolist())
    )

    for r in right:
        right_chart.add_yaxis(
            r,
            data[r].tolist(),
            yaxis_index=1,
            linestyle_opts=opts.LineStyleOpts(curve=360, type_='dotted'),
            label_opts=opts.LabelOpts(is_show=False),
        )

    left_chart.overlap(right_chart)
    return left_chart.render_notebook() if is_show else left_chart

In [10]:
week_day_n = data.groupby('week')['day'].apply(lambda x: x.nunique())

time_type = 'week'
uv_df = pd.DataFrame(data.groupby(time_type)['user_id'].apply(lambda x: x.nunique())).rename(columns={'user_id': 'uv'})
behaviour_df = data.groupby(time_type)[['view', 'cart', 'remove_from_cart', 'purchase', 'turnovers']].sum()
bar_title, bar_subtitle = f"每周的日均 PV/UV", '注：汇总方式 - 按周求和再除以当周天数；该月第 44 周仅 3 天，第 48 周仅 6 天。'
line_title, line_subtitle = f"每周的日均行为指标汇总", '注：汇总方式 - 按周求和再除以当周天数；该月第 44 周仅 3 天，第 48 周仅 6 天。'
uv_df['pv'] = behaviour_df['view']
uv_df = uv_df[['pv', 'uv']]
for c in uv_df.columns:
    uv_df[c] = round(uv_df[c] / week_day_n)
for c in behaviour_df.columns:
    behaviour_df[c] = round(behaviour_df[c] / week_day_n)
show_time_trend(uv_df, behaviour_df, bar_title, bar_subtitle, line_title, line_subtitle)

现象：47、48 周的各项日均指标均高于 44、45、46 周。

结论：推测 47、48 周有活动，用户活跃度高。

### 按星期分析

In [11]:
weekday_n = data.groupby('weekday')['week'].apply(lambda x: x.nunique())

time_type = 'weekday'
uv_df = pd.DataFrame(data.groupby(time_type)['user_id'].apply(lambda x: x.nunique())).rename(columns={'user_id': 'uv'})
behaviour_df = data.groupby(time_type)[['view', 'cart', 'remove_from_cart', 'purchase', 'turnovers']].sum()
bar_title, bar_subtitle = f"日均星期 PV/UV", '注：汇总方式 - 按星期求和再除以该星期次数；该月有两个星期四和星期五。'
line_title, line_subtitle = f"日均星期行为指标", '注：汇总方式 - 按星期求和再除以该星期次数；该月有两个星期和四星期五。'
uv_df['pv'] = behaviour_df['view']
uv_df = uv_df[['pv', 'uv']]
for c in uv_df.columns:
    uv_df[c] = round(uv_df[c] / weekday_n)
for c in behaviour_df.columns:
    behaviour_df[c] = round(behaviour_df[c] / weekday_n)
show_time_trend(uv_df, behaviour_df, bar_title, bar_subtitle, line_title, line_subtitle)

现象：周三、周四的各项日均指标明显高于其他。

结论：推测某个周三、周四有活动，用户活跃度高。

### 按日分析

In [12]:
time_type = 'day'
uv_df = pd.DataFrame(data.groupby(time_type)['user_id'].apply(lambda x: x.nunique())).rename(columns={'user_id': 'uv'})
behaviour_df = data.groupby(time_type)[['view', 'cart', 'remove_from_cart', 'purchase', 'turnovers']].sum()
bar_title, bar_subtitle = f"每日 PV/UV", '注：汇总方式 - 按日期求和'
line_title, line_subtitle = f"每日行为指标", '注：汇总方式 - 按日期求和'
uv_df['pv'] = behaviour_df['view']
uv_df = uv_df[['pv', 'uv']]
show_time_trend(uv_df, behaviour_df, bar_title, bar_subtitle, line_title, line_subtitle)

In [13]:
user_daily = pd.DataFrame(data.groupby('day')['user_id'].apply(lambda x: x.nunique()))
turnovers_day = pd.DataFrame(data.groupby('day')['turnovers'].sum())
user_daily = pd.concat([user_daily, turnovers_day], axis=1)
user_daily.columns = ['DAU', '日成交额']
user_daily['日客单价'] = user_daily['日成交额'] / user_daily['DAU']
user_daily.index.name = '日期'

double_y(user_daily, left=['日客单价'], right=['DAU', '日成交额'], title='日客单价变化')

In [14]:
# date_list = [f"2019-11-{i+1}" for i in range(30)]
date_list = [f"{i+1}" for i in range(30)]
dut_data = list()
dut = data.groupby(['day', 'user_id'])[['turnovers']].sum()
dut = dut[dut['turnovers'] > 0].reset_index().drop(columns=['user_id'])
for i in dut['day'].unique():
    dut_data.append(list(dut[dut['day'] == i]['turnovers']))

dut_kline = (
    Kline()
    .set_global_opts(
        title_opts=opts.TitleOpts(title='用户日成交额'),
        xaxis_opts=opts.AxisOpts(name='日期'),
        yaxis_opts=opts.AxisOpts(name='用户日成交额'),
    )
    .add_xaxis(date_list)
    .add_yaxis('', dut_data)
)
dut_kline.render_notebook()

In [15]:
# # date_list = [f"2019-11-{i+1}" for i in range(30)]
# date_list = [f"{i+1}" for i in range(30)]
# dut_data = list()
# dut = data.groupby(['day', 'user_id'])[['turnovers']].sum()
# dut = dut[dut['turnovers'] > 0].reset_index().drop(columns=['user_id'])
# for i in dut['day'].unique():
#     dut_data.append(list(dut[dut['day'] == i]['turnovers']))

    
# dut_boxplot = Boxplot()
# dut_boxplot.set_global_opts(
#     title_opts=opts.TitleOpts(title='用户日成交额'),
#     xaxis_opts=opts.AxisOpts(name='日期'),
#     yaxis_opts=opts.AxisOpts(name='用户日成交额'),
# )
# dut_boxplot.add_xaxis(date_list)
# dut_boxplot.add_yaxis("", dut_boxplot.prepare_data(dut_data))
# dut_boxplot.render_notebook()

In [16]:
weekday_dict = dict(zip([i for i in range(7)], ['星期日', '星期一', '星期二', '星期三', '星期四', '星期五', '星期六']))
day_weekday = data[['weekday', 'day']].drop_duplicates().reset_index(drop=True)
day_weekday['weekday'] = day_weekday['weekday'].map(weekday_dict)
day_weekday = pd.merge(day_weekday, data[['day', 'week']].drop_duplicates(), how='left', on='day')
print("【日期-星期对应表】")
day_weekday.set_index('day').T

【日期-星期对应表】


day       1    2    3    4    5    6    7    8    9    10   11   12   13   14  \
weekday  星期四  星期五  星期六  星期日  星期一  星期二  星期三  星期四  星期五  星期六  星期日  星期一  星期二  星期三   
week      44   44   44   45   45   45   45   45   45   45   46   46   46   46   

day       15   16   17   18   19   20   21   22   23   24   25   26   27   28  \
weekday  星期四  星期五  星期六  星期日  星期一  星期二  星期三  星期四  星期五  星期六  星期日  星期一  星期二  星期三   
week      46   46   46   47   47   47   47   47   47   47   48   48   48   48   

day       29   30  
weekday  星期四  星期五  
week      48   48

现象：
- 有三个日期区间的各项指标高于均值，按超出程度依次为：
  - 21 号 ~ 24 号：47 周的周三至周六
  - 28 号 ~ 30号：48 周的周三至周五
  - 7 号~ 8 号：45 周的周三至周四

- 以上区间后的一两天各项指标都会有大幅的下跌，再逐步恢复正常。
  - 9 号各项指标环比骤降，随后 9 号 ~10 号且明显低于日常，再逐步回升，于 11 日恢复正常
  - 25 号各项指标环比骤降且明显低于日常，再逐步回升，于 11 日恢复正常

结论：
- 应证了之前每周指标和星期指标的结论，推测以上日期区间都推出了活动，且活动力度按以上顺序依此递减。
- 活动后的一两天会陷入疲软期，各项指标明显低于日常再逐渐恢复。活动之间间隔较短可能导致活动间隔期间各项指标持续疲软，直至活动再大幅改善。

建议：
- 根据每次活动后的活跃度下降情况，合理控制活动的周期与频率。
- 根据每次活动的 ROI，吸取 ROI 高的活动的优点，总结 ROI 低的活动里需要改进和避免的点。

### 按时段分析

In [17]:
time_type = 'hour'
uv_df = pd.DataFrame(data.groupby(time_type)['user_id'].apply(lambda x: x.nunique())).rename(columns={'user_id': 'uv'})
behaviour_df = data.groupby(time_type)[['view', 'cart', 'remove_from_cart', 'purchase', 'turnovers']].sum()
bar_title, bar_subtitle = f"各时段 PV/UV", '注：汇总方式 - 按时段求和'
line_title, line_subtitle = f"各时段行为指标", '注：汇总方式 - 按时段求和'
uv_df['pv'] = behaviour_df['view']
uv_df = uv_df[['pv', 'uv']]
show_time_trend(uv_df, behaviour_df, bar_title, bar_subtitle, line_title, line_subtitle)

In [18]:
behaviour_df['客单价'] = np.round(behaviour_df['turnovers'] / uv_df['uv'], 4)
behaviour_df['浏览-购买转化率'] = np.round(behaviour_df['purchase'] / behaviour_df['view'], 4)
behaviour_df['浏览-加购转化率'] = np.round(behaviour_df['cart'] / behaviour_df['view'], 4)
behaviour_df['加购-移除转化率'] = np.round(behaviour_df['remove_from_cart'] / behaviour_df['cart'], 4)
behaviour_df.index.name = '时段'

double_y(
    behaviour_df,
    left=['客单价'],
    right=['浏览-购买转化率', '浏览-加购转化率', '加购-移除转化率'],
    title='时段客单价与转化率',
)

现象：
1. 19 时为夜高峰，同时当日最高峰；
2. 20 时起持续下降，至夜里 0、1 时低至当日最低；
3. 2 时起逐步回升，于 11、12 时达到日高峰；
4. 13 时至 15 时小幅回落，16 时起回升，至 19 时达到夜高峰。
5. 11 时至 13 时和 19 时至 22 时的客单价高于各时段的客单价均值。
6. 除去 0 时至 6 时这一特殊时段，其余时段的各项转化率变化幅度小，波动不明显。

结论：
- 一天之中存在两个用户活跃高峰，分别为：
    - 日高峰：中午 11~12 时左右
    - 夜高峰：晚上 18~20 时左右
- 对比各时段客单价，19 时至 22 时用户质量最高，其次是 11 时至 13 时。

建议：
- 注重日高峰和夜高峰（尤其是夜高峰）的运营：
    - 在该时段进行活动和商品推送，如：新品上线、商品限时折扣。
    - 在该时段限时价格优惠，促进用户购买，如：如促销或库存类商品优惠券。

## 用户分析

### 用户行为统计信息

用户汇总后的行为统计信息：

In [19]:
user_gb.describe()

view          cart  remove_from_cart      purchase  \
count 368232.000000 368232.000000     368232.000000 368232.000000   
mean       5.633742      3.499742          1.901844      0.875136   
std       25.864980     15.198215         12.028245      4.564265   
min        0.000000      0.000000          0.000000      0.000000   
25%        1.000000      0.000000          0.000000      0.000000   
50%        1.000000      0.000000          0.000000      0.000000   
75%        3.000000      1.000000          0.000000      0.000000   
max     9335.000000   1455.000000       1547.000000    438.000000   

          turnovers  
count 368232.000000  
mean       4.157246  
std       21.640169  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max     2715.870000

In [20]:
# ### 整体情况
# per_capita_behaviour = round(data[['view', 'cart', 'remove_from_cart', 'purchase', 'turnovers']].sum() / data['user_id'].nunique(), 2)
# per_capita_behaviour = [list(per_capita_behaviour.values)]
# # per_capita_behaviour = [list(z) for z in zip(per_capita_behaviour.index.tolist(), per_capita_behaviour.tolist())]
# headers = ['人均浏览', '人均加购', '人均移出购物车', '人均购买次数', '人均成交额']

# per_capita_behaviour_table = (
#     Table()
#     .add(headers, per_capita_behaviour)
# )
# per_capita_behaviour_table.render_notebook()

根据均值-方差方式（超出均值 3 个标准差的区间视为异常值）去除异常值后的统计信息：

In [21]:
def u_3sigma(data):
    data_outliars = pd.DataFrame()
    for f in data.columns:
        u, sigma = data[f].mean(), data[f].std()
        lower, upper = u - 3 * sigma, u + 3 * sigma
        data_outliars[f] = data[f].apply(lambda x: False if x >= lower and x <= upper else True)
    return data_outliars


user_outliars = u_3sigma(user_gb)
user_clear = user_gb[~user_outliars.any(axis=1)]
user_clear
user_clear.describe()

view          cart  remove_from_cart      purchase  \
count 355432.000000 355432.000000     355432.000000 355432.000000   
mean       3.779792      1.767925          0.765010      0.316933   
std        7.345889      5.180515          3.182895      1.512991   
min        0.000000      0.000000          0.000000      0.000000   
25%        1.000000      0.000000          0.000000      0.000000   
50%        1.000000      0.000000          0.000000      0.000000   
75%        3.000000      0.000000          0.000000      0.000000   
max       83.000000     49.000000         37.000000     14.000000   

          turnovers  
count 355432.000000  
mean       1.607671  
std        7.436077  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max       69.040000

### 用户活跃

指标口径：
- 日新增用户：当日是该月第一次登陆的用户
- 日流失用户：当日是该月最后一次登录的用户

In [22]:
user_day_n = data.groupby('user_id')[['day']].agg(['min', 'max'])
user_day_n.columns = ['new', 'loss']
user_day_n = pd.DataFrame(
    [user_day_n['new'].value_counts().sort_index(),
    user_day_n['loss'].value_counts().sort_index()]
).T

day_new_line = (
    Line()
    .set_global_opts(
        title_opts=opts.TitleOpts(
            title='当月日新增/日流失用户数'
        ),
        tooltip_opts=opts.TooltipOpts(
            is_show=True, trigger="axis", axis_pointer_type="cross"
        ),
        xaxis_opts=opts.AxisOpts(
            type_="category",
            axispointer_opts=opts.AxisPointerOpts(is_show=True, type_="shadow"),
        ),
    )
    .add_xaxis(user_day_n.index.astype('str').tolist())
    .add_yaxis('new', user_day_n['new'].tolist())
    .add_yaxis('loss', user_day_n['loss'].tolist())
    .set_series_opts(label_opts=opts.LabelOpts(is_show=False))
)
day_new_line.render_notebook()

现象：活动期间用户活跃度明显增高。

结论：利用活动期，同时做好拉新与召回。

建议：
- 拉新：
    1. 以老带新，邀请双方优惠
    2. 通过社交媒体、广告增大曝光量，吸引新用户
- 召回：
    1. 针对流失用户（长周期内未使用）：在重大活动或改版的时候，以邮箱、电话、推广的方式召回
    2. 针对沉默用户（长周期内使用过，短周期内未使用）：通过内容推荐、定期活动推送召回

### 用户留存

指标口径：
- 第 N 日留存率：当日为该月第一次登录的用户在之后的第 N 日仍然登录的比例

In [23]:
user_new_day = data.groupby('user_id')[['day']].min()
user_new_day.columns = ['first_day']
user_day = pd.merge(user_new_day, data[['user_id', 'day']], on='user_id', how='left')
user_day['nday'] = user_day['day'] - user_day['first_day']
day_retention_n = user_day[['first_day', 'nday', 'user_id']].groupby(['first_day', 'nday']).count()
day_retention_n = day_retention_n.reset_index()

retention_table = day_retention_n.pivot_table(index=['first_day'], columns=['nday'])
retention_table.columns = [i for i in range(retention_table.shape[1])]
first_day_n = retention_table[0]
del retention_table[0]
for col in retention_table.columns:
    retention_table[col] = np.round(retention_table[col] / first_day_n, 4)
    
# rentention_rate = retention_table[[1, 3, 7]]
# rate_columns = [f'{col} 日留存率' for col in rentention_rate.columns]
# rentention_rate.columns = rate_columns
# rentention_rate['turnovers'] = data.groupby('day')['turnovers'].sum()
# double_y(rentention_rate, left=rate_columns, right=['turnovers'], title='用户留存')

retention_table = retention_table.iloc[::-1].T
retention_values = [[i, j, val] for i, row in enumerate(retention_table.values) for j, val in enumerate(row)]
retention_heatmap = (
    HeatMap()
    .set_global_opts(
        title_opts=opts.TitleOpts(title='用户留存热力图'),
        xaxis_opts=opts.AxisOpts(
            name='第 N 日留存',
            position='top'
        ),
        yaxis_opts=opts.AxisOpts(
            name='日期',
            name_location='start',
            axislabel_opts=opts.LabelOpts(font_size=9),
        ),
        visualmap_opts=opts.VisualMapOpts(
            max_=0.25,
            is_calculable=True,
            pos_top='middle',
            pos_left='right',
            precision=2,
        ),
    )
    .add_xaxis(retention_table.index.astype(str).tolist())
    .add_yaxis("第 N 日留存", retention_table.columns.astype(str).map(lambda x: f"2019-11-{x}").tolist(), retention_values)
)
retention_heatmap.render_notebook()

现象：留存率变化不稳定，无明显规律，在活动期间有明显波动。

结论：对于电商，大多数用户只在有购物需求时才会使用，不像社交/游戏产品具有日常高频的使用需求，因而难有稳定且高的留存率。

### 用户登录天数

In [24]:
# 访问次数
user_view_cnt = data[['user_id', 'view', 'turnovers']].groupby('user_id').sum()
# 访问天数
user_login_days = data[['user_id', 'day']].drop_duplicates().groupby('user_id').count()
user_login_days.columns = ['login_days']
user_stickiness = pd.concat([user_view_cnt, user_login_days], axis=1)
user_stickiness['average_daily_view'] = user_stickiness['view'] / user_stickiness['login_days']
# user_stickiness
login_days_df = user_stickiness.groupby('login_days')[['turnovers']].mean()
login_days_df.index.name = '累计登陆天数'
login_days_df.columns = ['平均成交额']
login_days_df['用户比例'] = user_stickiness['login_days'].value_counts(normalize=True).sort_index()
login_days_df['平均日均访问量'] = user_stickiness.groupby('login_days')[['average_daily_view']].mean()
double_y(login_days_df, left=['用户比例'], right=['平均成交额', '平均日均访问量'], title='用户累计登陆天数分布')

现象：
- 随着登陆天数的增大，相应的用户数量迅速降低：当月 78% 的用户仅登录 1 天，12% 的用户仅登录 2 天，仅有 10 % 的用户登陆天数大于 3 天
- 整体上，登陆天数越大，用户的日均访问量与成交额越高

结论：
- 用户粘性需要提高
- 用户粘性越大，用户产生的价值（流量、收益）越高

建议：针对登陆天数少的用户采取措施进行优化。除去电商本就不如即时通讯类具有日常高频的使用需求，导致用户登陆天数少的原因还有用户使用体验不佳、对内容不感冒等。因此优化可以有：
1. 推出日常任务体系（如：签到、浏览可得积分并换取优惠券）
2. 针对提优化界面和操作，提高用户的好感
3. 提高推荐系统精准度，适时推送内容

### 用户购买情况

In [25]:
def draw_pareto(data, tittle='帕累托图', x_name='', is_show=True):
    '''
    Draw pareto.
    '''
    data[f'{data.columns[0]}-累计占比'] = np.round(data.values[:, 0].cumsum() / data.values[:, 0].sum(), 2)
    
    bar = (
        Bar()
        .set_global_opts(
            title_opts=opts.TitleOpts(title=tittle),
            tooltip_opts=opts.TooltipOpts(
                is_show=True, trigger="axis", axis_pointer_type="cross"
            ),
            xaxis_opts=opts.AxisOpts(
                name=x_name,
                name_location='center',
                name_gap=30,
                type_="category",
                axispointer_opts=opts.AxisPointerOpts(is_show=True, type_="shadow"),
            ),
            yaxis_opts=opts.AxisOpts(
                name=data.columns[0],
                type_="value",
                axislabel_opts=opts.LabelOpts(formatter="{value}"),
                axistick_opts=opts.AxisTickOpts(is_show=True),
            ),
            legend_opts=opts.LegendOpts(
                type_="scroll",
            ),
        )
        .extend_axis(
            yaxis=opts.AxisOpts(
                name=data.columns[1],
                type_="value",
                min_=0,
                max_=1.1,
                axislabel_opts=opts.LabelOpts(formatter="{value}"),
            )
        )
        .add_xaxis(data.index.astype(str).tolist())
        .add_yaxis(
            data.columns[0], 
            data.iloc[:, 0].tolist(),
            z=0
        )
        .set_series_opts(label_opts=opts.LabelOpts(is_show=False))
    )

    cumsum_line = (
        Line()
        .add_xaxis(data.index.astype(str).tolist())
        .add_yaxis(
            data.columns[1], 
            data.iloc[:, 1].tolist(),
            yaxis_index=1,
            linestyle_opts=opts.LineStyleOpts(width=1),
            markline_opts=opts.MarkLineOpts(
                data=[opts.MarkLineItem(y=0.8)],
                linestyle_opts=opts.LineStyleOpts(
                    width=1,
                    type_='dashed',
                    opacity=0.8,
                ),
                symbol='none',
                label_opts=opts.LabelOpts(is_show=False),
#                 label_opts=opts.LabelOpts(
#                     formatter='80% 水平线',
#                     position='left',
#                 ),
            ),
            z=1
        )
        .set_series_opts(label_opts=opts.LabelOpts(is_show=False))
    )
    
    return bar.overlap(cumsum_line).render_notebook() if is_show else bar.overlap(cumsum_line)

In [26]:
turnovers_vc = pd.cut(
    user_gb[user_gb['turnovers'] > 0]['turnovers'],
    bins=[i for i in range(0, int(np.max(user_gb['turnovers'])) + 50, 50)],
).value_counts().sort_index()
turnovers_vc = pd.DataFrame(turnovers_vc).rename(columns={'turnovers': '用户数'})
# turnovers_vc
draw_pareto(pd.DataFrame(turnovers_vc), x_name='消费金额', tittle='消费金额分布')

现象：
- 70% 的客户消费金额在 (0, 50] 的区间内
- 20% 的客户消费金额在 (50, 100] 的区间内
- 近 100% 的用户消费金额在 (0, 300] 的区间内

建议：发掘客户潜力，提高消费金额。

In [27]:
purchase_vc = user_gb[user_gb['purchase'] > 0]['purchase'].value_counts().sort_index()
purchase_vc.index = purchase_vc.index.astype(int)
purchase_vc = pd.DataFrame(purchase_vc).rename(columns={'purchase': '用户数'})
# purchase_vc
draw_pareto(purchase_vc, tittle='购买次数分布', x_name='购买次数')

现象：
- 用户当月最大购买次数为 438
- 50% 的客户当月购买次数在 ≤ 6
- 80% 的客户当月购买次数在 ≤ 15

结论：客户复购频率较高。

建议：保持用户粘性。

In [28]:
user_purchase = data[data['purchase'] == 1][['user_id', 'day']].drop_duplicates().reset_index(drop=True)
user_merge = pd.merge(left=user_purchase, right=user_purchase, how='left', on='user_id')
user_merge = user_merge[user_merge['day_y'] > user_merge['day_x']]
user_repurchase = user_merge.groupby(['user_id', 'day_x']).first()
user_repurchase = user_repurchase.reset_index()
user_repurchase['ndays'] = user_repurchase['day_y'] - user_repurchase['day_x']
# user_repurchase
repurchase_days = user_repurchase['ndays'].value_counts().sort_index()
repurchase_days = pd.DataFrame(repurchase_days).rename(columns={'ndays': '复购次数'})
# repurchase_days
draw_pareto(repurchase_days, tittle='复购间隔天数分布', x_name='复购间隔天数')

现象：整体上，随着购后间隔天数增大，发生复购的次数越来越少。

结论：购后的间隔天数越小，复购的概率越高。

建议：在用户发生购买行为后的几天里，推荐与购买商品相关的其它商品。

### 行为路径分析

In [29]:
up_path = data.groupby(['user_id', 'product_id'])['event_type'].apply(lambda x: ' -> '.join(x))
up_path = pd.DataFrame(up_path)
# up_path
path_distribution = up_path[['event_type']].value_counts(normalize=True)
# path_distribution

path_pie = (
    Pie()
    .add(
        '',
        [list(z) for z in zip(path_distribution[path_distribution > 0.01].index.tolist(), path_distribution[path_distribution > 0.01].values.tolist())],
        radius=[100, 180],
#         rosetype="radius",
    )
    .set_global_opts(
        title_opts=opts.TitleOpts(
            title='用户-物品行为路径占比',
            subtitle='注：仅展示占比 ≥1% 的行为路径',
        ),
        legend_opts=opts.LegendOpts(is_show=False)
    )
    .set_series_opts(label_opts=opts.LabelOpts(formatter='{b}: {d}%'))
)
path_pie.render_notebook()

In [30]:
# user_product_gb = data.groupby(['user_id', 'product_id'])['view', 'cart', 'remove_from_cart', 'purchase', 'turnovers'].sum()
# behaviour_path = user_product_gb[[col for col in user_product_gb.columns if col != 'turnovers']]
# behaviour_path_copy = behaviour_path.copy()
# for b in behaviour_path_copy.columns:
#     behaviour_path_copy[b] = behaviour_path_copy[b].apply(lambda x: True if x > 0 else False)
# behaviour_path_copy = behaviour_path_copy.reset_index()
# path_cnt = behaviour_path_copy.groupby(['view', 'cart', 'remove_from_cart', 'purchase']).count()
# del path_cnt['user_id']
# path_cnt.columns = ['cnt']
# # path_cnt
# path = path_cnt.reset_index()
# for col in [col for col in path.columns if col != 'cnt']:
#     path[col] = path[col].apply(lambda x: col if x else '')
# # path
# path = path.sort_values(by=[col for col in path.columns if col != 'cnt'], ascending=False, ignore_index=True)
# path_name = list()
# for i in range(len(path)):
#     path_name.append('→'.join(b for b in path.loc[i][:-1] if b != ''))
# path['path'] = path_name
# # path
# path = path[['path', 'cnt']].set_index('path').sort_values('cnt', ascending=False)
# # path

# path_pie = (
#     Pie()
#     .add(
#         '',
#         [list(z) for z in zip(path.index.tolist(), path['cnt'].tolist())],
#         radius=[100, 180],
# #         rosetype="radius",
#     )
#     .set_global_opts(
#         title_opts=opts.TitleOpts('用户-物品行为路径占比'),
#         legend_opts=opts.LegendOpts(is_show=False)
#     )
#     .set_series_opts(label_opts=opts.LabelOpts(formatter='{b}: {d}%'))
# )
# path_pie.render_notebook()

现象：

关注占比大于 5% 的行为路径，按占比从高到低依次为：

|行为|占比(%)|
|-|-|
|一次浏览|48.35|
|仅加购|13.59|
|加购后移除|10.56|
|加购后购买|6.82|
|二次浏览|6.37|
|移除购物车|5.01|

结论：
- “浏览”转化率低：“一次浏览”、“二次浏览”、“三次浏览”分别占比48.3% 、6.37%、1.45%，都没有进一步的行为转化
- “加购-购买”转化率低：没有操作 > 移除 > 购买，且“移除购物车”（5.01%）占比明显高于“购买”（1.28%），可见，用户“加购”并不一定代表较高的购买欲望，也可能是商品收藏或价格观察。

建议：
- 提高推荐系统精准度，使商品的浏览推荐更符合用户需求。
- 推出收藏功能，既完善用户使用的功能性，又便于我们分析用户的意图。
- 当加购后的商品发生信息变更（活动、降价）或久未操作，及时提醒用户。例如：通过手机、邮箱等联系方式推送

## 客户分析

构建 RFM 模型，对用户进行分层，方便跟踪各层用户的特征及变化情况，以针对性运营提高效率。

目标指标：
- Recency: 最近一次消费时间间隔
- Frequency：消费频率
- Monetary：消费金额

时间区间：11-01 ~ 11-30，以 12-01 为当前时间。

划分方式：以均值作为划分标准，将 RFM 三个指标数据转化为高、低两个等级变量，根据用户每项指标等级，划分为 8 类具不同消费特征的群体，由此直观反映各类客户的价值和重要性，并可根据相应特征采取针对性策略，可操作性强。

|R|F|M|划分群体类型|特点|
|-|-|-|-|-|
|高|高|高|重要价值客户|高端客户，各项指标都优于平均，质量最优|
|低|高|高|重要保持客户|最近购买日期早于平均最近购买日期，存在流失的可能性|
|高|低|高|重要发展客户|消费次数低于均值，用户粘性不高|
|低|低|高|重要挽留客户|消费次数低于均值和最近购买日期早于平均最近购买日期，流失风险大|
|高|高|低|一般价值客户|近期消费、频率高于平均，但消费金额低于平均|
|低|高|低|一般保持客户|频率高于平均、消费金额低于平均，近期无消费，存在流失的可能性|
|高|低|低|一般发展客户|近期消费，频率和消费金额均低于平均，可能是试水新用户或粘性不高的老用户|
|低|低|低|一般挽留客户|各项指标都低于平均|

注：
1. 由于消费金额 M 是衡量用户价值更重要的属性，故根据消费高金额 M 的高低又划分为重要客户和一般客户。
2. 由于存在大量仅浏览用户，故本次划分只选取**本月有过消费行为的用户**为目标划分群体。

In [31]:
# 1. 统计消费用户 rfm 指标
user = data[['user_id']].groupby('user_id').count()
recently_purchase_day = data[data['purchase'] == 1][['user_id', 'day']].groupby('user_id').apply(lambda x: x.max())
recently_purchase_day = recently_purchase_day.drop(columns=['user_id'])
user_recently = pd.merge(left=user, right=recently_purchase_day, on='user_id', how='left')
user_recently = user_recently.fillna(value=0)
user_frequency = data[['user_id', 'purchase']].groupby('user_id').sum()
user_monetary = data[['user_id', 'turnovers']].groupby('user_id').sum()
user_rfm = pd.concat([user_recently, user_frequency, user_monetary], axis=1)
user_rfm.columns = ['Recently','Frequency', 'Monetary']
consumer_rfm = user_rfm[user_rfm['Frequency'] > 0]

# 2. 划分 rfm 等级变量
rfm_divide = {col: consumer_rfm[col].mean() for col in consumer_rfm.columns}  # 取 rfm 中位数
consumer_rfm = pd.DataFrame(consumer_rfm)

for col in consumer_rfm:
    '''等级变量'''
    consumer_rfm[f"_{col}"] = consumer_rfm[col].apply(lambda x: 1 if x > rfm_divide[col] else 0)

# 3. 根据 rfm 等级变量打标签
consumer_rfm['code'] = 100 * consumer_rfm['_Recently'] + 10 * consumer_rfm['_Frequency'] + consumer_rfm['_Monetary']  # 编码
rfm_code2label = {
    111: '重要价值客户',
    11: '重要保持客户',
    101: '重要发展客户',
    1: '重要挽留客户',
    110: '一般价值客户',
    10: '一般保持客户',
    100: '一般发展客户',
    0: '一般挽留客户',
}
consumer_rfm['label'] = consumer_rfm['code'].apply(lambda x: rfm_code2label[x])  # 编码转标签

# 3. 分层效果检验
# 3.1 各类用户占比
labels_order = ['重要价值客户', '重要保持客户', '重要发展客户', '重要挽留客户', '一般价值客户', '一般保持客户', '一般发展客户', '一般挽留客户']
label_vc = consumer_rfm['label'].value_counts().reindex(labels_order)

rfm_pie = (
    Pie()
    .add(
        '',
        [list(z) for z in zip(label_vc.index.tolist(), label_vc.values.tolist())],
        radius=[80, 180],
        rosetype="radius", #玫瑰饼图
    )
    .set_global_opts(
        title_opts=opts.TitleOpts('RFM 分层效果'),
        legend_opts=opts.LegendOpts(is_show=False)
    )
    .set_series_opts(label_opts=opts.LabelOpts(formatter='{b}: {d}%'))
)

# 3.2 各类用户 rfm 雷达图
label_mean = consumer_rfm.groupby('label')[['Recently', 'Frequency', 'Monetary']].mean()
label_mean = label_mean.reindex(labels_order)
colors = ['#D53A35', '#2F4554', '#61A0A8', '#D48265', '#749F83', '#CA8622', '#CA8622', '#6E7074']
# 数据做归一化处理
from sklearn.preprocessing import MinMaxScaler
label_mean_mms = MinMaxScaler().fit_transform(label_mean)

label_mean_radar = (
    Radar()
    .add_schema(
        schema=[opts.RadarIndicatorItem(name=col) for col in label_mean.columns]
    )
    .set_global_opts(
        title_opts=opts.TitleOpts("各类客户 RFM 雷达图"),
        legend_opts=opts.LegendOpts(pos_bottom='15%'),
    )
)

for i, v in enumerate(label_mean.index):
    label_mean_radar.add(v, [label_mean.loc[v].tolist()], color=colors[i])
    label_mean_radar.set_series_opts(label_opts=opts.LabelOpts(is_show=False))
    
label_monetary = pd.DataFrame(consumer_rfm.groupby('label')['Monetary'].sum()).reindex(labels_order)
label_monetary.columns = ['消费金额']
pareto = draw_pareto(label_monetary, x_name='客户类型', is_show=False)
    
rfm_page = (
    Page()
    .add(rfm_pie, label_mean_radar, pareto)
)
rfm_page.render_notebook()

现象：
- 客户按比例由高到低依次为：一般发展客户(30.82%)，一般挽留客户(27.19%)，重要价值客户(13.94%)，重要保持客户(7.27%)，一般价值客户(6.36%)，重要发展客户(5.27%)，重要挽留客户(4.59%)，一般保持客户(4.11%)。
- 重要客户占整体比例 30.5%，贡献整体销售额的 66%。其中，重要价值用户占整体比例 13.94%，贡献整体销售额的 33%。

结论：高端客户和低端客户两级分化，中间层次用户较少，低端客户有较大开发空间。

建议：
- 经营好高端客户，保持高端客户的粘性。例如：
    - '重要价值客户'： VIP 优待 -- 提供更优质的售时售后服务，或进行不定期福利回馈。
    - '重要保持客户'：及时唤醒 -- 采取措施重新唤醒，促进用户复购。若存有联系方式，而可通过固定周期未购买进行消息通知。
    - '重要发展客户' 和 '重要挽留客户'：提高客户粘性 -- 增加对这类客户的曝光量和吸引力。前者可以及时推送优惠活动，后者可以通过调研客户消费需求和满意度调查，对能够改进的地方进行完善。
- 发充分发掘低端客户，使其向更高层次转化。例如：针对这部分客户的进行消费需求调研和满意度调查，对能够改进的地方进行完善。

## 品牌分析

In [32]:
brand_gb_fun = {
    'view': np.sum, 
    'cart': np.sum, 
    'remove_from_cart': np.sum, 
    'purchase': np.sum, 
    'turnovers': np.sum,
    'price': [np.median, np.mean],  # [np.min, np.max, np.median, np.mean]
    'product_id': lambda x: x.nunique()
}
brand_gb = data.groupby('brand').agg(brand_gb_fun)
# brand_gb.columns = ['view', 'cart', 'remove_from_cart', 'purchase', 'turnovers', 'min_price', 'max_price', 'median_price', 'mean_price', 'products_n']
brand_gb.columns = ['view', 'cart', 'remove_from_cart', 'purchase', 'turnovers', 'median_price', 'mean_price', 'products_n']
brand_gb = brand_gb.drop(index=['unknown'])  # 删除未知品牌
# brand_gb['avg_view'] = np.round(brand_gb['view'] / brand_gb['products_n'], 2)
# brand_gb['avg_cart'] = np.round(brand_gb['cart'] / brand_gb['products_n'], 2)
# brand_gb['avg_purchase'] = np.round(brand_gb['purchase'] / brand_gb['products_n'], 2)
brand_gb['purchase/view'] = np.round(brand_gb['products_n'] / brand_gb['view'], 2)
brand_gb['purchase/view'] = brand_gb['purchase/view'].fillna(0)
brand_gb['turnovers/purchase'] = np.round(brand_gb['turnovers'] / brand_gb['purchase'], 2)
brand_gb['turnovers/purchase'] = brand_gb['turnovers/purchase'].fillna(0)

In [33]:
brandn_wc_data = [(brand, n) for brand, n in zip(brand_gb.index.tolist(), brand_gb['products_n'].tolist())]
brand_wc = (
    WordCloud()
    .add('品牌商品数量', brandn_wc_data)
    .set_global_opts(title_opts=opts.TitleOpts('品牌商品数量'))
)
brand_wc.render_notebook()

现象：三个品牌的商品数量大于 1000 -- masura(1562)、irisk(1325)、runail(1274)，数量级明显大于其它品牌（freedecor(840) 商品数量排名第四）。

In [34]:
brand_turnovers = brand_gb[brand_gb['turnovers'] > 0][['turnovers']].sort_values(by=['turnovers'], ascending=False)
brand_turnovers.columns = ['品牌成交额']
brand_turnovers['rank'] = brand_turnovers['品牌成交额'].rank(method='first', ascending=False)
brand_turnovers['rank'] = brand_turnovers['rank'].astype(int)
brand_turnovers = brand_turnovers.reset_index()
brand_turnovers['brand'] = brand_turnovers['rank'].astype(str) + '.' + brand_turnovers['brand'].astype(str)
brand_turnovers = brand_turnovers.set_index('brand')
del brand_turnovers['rank']
# brand_turnovers
draw_pareto(brand_turnovers, tittle='品牌成交额帕累托图')

现象：成交额排名前 36（前 15%）的品牌贡献了 80% 的成交额，接近“二八法则”。

In [35]:
name = brand_gb.index.tolist()
bvpt = brand_gb[['view', 'purchase', 'turnovers']]

formatter_js = '''
    function(params){
        return '品牌：' + params.value[2]
        + '<br/>流量：' + params.value[0]
        + '<br/>销量：' + params.value[1]
        + '<br/>成交额：' + params.value[3];
    }
    '''

symbol_js = """
    function (data) {return Math.sqrt(data[3] / 88);}
    """

pt_scatter = (
    Scatter()
    .add_xaxis(bvpt['view'])
    .add_yaxis(
        "",
        [list(z) for z in zip(bvpt['purchase'], name, bvpt['turnovers'])],
        symbol_size=JsCode(symbol_js),
        label_opts=opts.LabelOpts(
            is_show=False,
            formatter=JsCode(
                "function(params){return params.value[3];}"
            ),
        ),
    )
    .set_global_opts(
        title_opts=opts.TitleOpts(
            title="品牌流量与销量矩阵",
            subtitle='注：点的大小与成交额正相关',
            pos_left='center',
        ),
        xaxis_opts=opts.AxisOpts(name='流量', type_="value"),
        yaxis_opts=opts.AxisOpts(name='销量', type_="value"),
#         visualmap_opts=opts.VisualMapOpts(
#             type_="color",
#             min_=0,
#             max_=np.max(bvpt['purchase']),
#             dimension=0,
#         ),
#         legend_opts=opts.LegendOpts(is_show=True),
#         toolbox_opts=opts.ToolboxOpts(is_show=True),
        tooltip_opts=opts.TooltipOpts(
            formatter=JsCode(formatter_js),
        ),
        datazoom_opts=opts.DataZoomOpts(is_show=True),
    )
    .set_series_opts(
        markline_opts=opts.MarkLineOpts(
            is_silent=True,
            data=[
                opts.MarkLineItem(
                    name="流量均值",
                    type_='value',
#                     x=round(np.mean(bvpt['view']), 0),
                    x=5020,
                ),
                opts.MarkLineItem(
                    name="销量均值",
                    type_='value',
#                     y=round(np.mean(bvpt['purchase']), 0),
                    y=772,
                ),
            ],
            symbol='none',
            label_opts=opts.LabelOpts(
                is_show=True,
                formatter="均值:{c}",
            ),
        ),
    )
)
# pt_scatter.render_notebook()

level = {
    11: '高流量-高销量',
    10: '高流量-低销量',
    1: '低流量-高销量',
    0: '低流量-低销量',
}

for c in ['view', 'purchase']:
    bvpt[f"{c}_level"] = (bvpt[c] > bvpt[c].mean()).astype(int)
bvpt['level'] = bvpt['view_level'] * 10 + bvpt['purchase_level']
bvpt['level'] = bvpt['level'].map(level)
# bvpt
level_vc = bvpt['level'].value_counts().sort_values().reindex(level.values())
# level_vc
level_pie = (
    Pie()
    .add(
        '',
        [list(z) for z in zip(level_vc.index.tolist(), level_vc.values.tolist())],
        radius=[80, 180],
    )
    .set_global_opts(
        title_opts=opts.TitleOpts(
            title='品牌矩阵分布',
            pos_left='center',
        ),
        legend_opts=opts.LegendOpts(is_show=False)
    )
    .set_series_opts(label_opts=opts.LabelOpts(formatter='{b}: {d}%'))
)
level_pie.render_notebook()

page = (
    Page()
    .add(pt_scatter, level_pie)
)
page.render_notebook()

现象：
1. 大体上，流量越高的品牌，销量和成交额也越高
2. 存在销量高但是成交额低的品牌
3. 根据品牌浏览量与成交量的**均值**将品牌划分至四象限：

    - 高流量，高销量：占比 15.9%
    - 高流量，低销量：占比 4.6%
    - 低流量，高销量：占比 1.26%
    - 低流量，低销量：占比 78.25%

结论：

- 流量与销量、成交额正相关

建议：合理控制定价，仅靠高曝光、高销量无法实现高成交额的目标。

In [36]:
# brand_turnovers_top10 = bvpt.sort_values('turnovers')[-10:]

# brand_turnovers_top10_bar = (
#     Bar()
#     .add_xaxis(brand_turnovers_top10.index.tolist())
#     .add_yaxis('成交额', brand_turnovers_top10['turnovers'].tolist())
#     .add_yaxis('浏览量', brand_turnovers_top10['view'].tolist())
#     .set_global_opts(title_opts=opts.TitleOpts('成交额最高的 10 个品牌'))
#     .set_series_opts(label_opts=opts.LabelOpts(position='inside', color='white'))
#     .reversal_axis()
# )
# brand_turnovers_top10_bar.render_notebook()